src/kademlia/node.cpp

```diff
@@ -1191,28 +1190,60 @@ void node::write_nodes_entries(sha1_hash const& info_hash
 		if (wanted_node == m_nodes.end()) continue;
 		std::vector<node_entry> n;
 		wanted_node->second->m_table.find_node(info_hash, n, 0);
-		r[wanted_node->second->protocol_nodes_key()] = write_nodes_entry(n);
+		r[wanted_node->second->protocol().nodes_key()] = write_nodes_entry(n);
 	}
 }
 
-node::protocol_descriptor const& node::map_protocol_to_descriptor(udp protocol)
```

Comment:
I think this approach of creating the instances and picking the right one kind of makes sense. your version is simpler in the sense that there's only one test and then it picks based on that, but it's more complex in the sense that you require the wrapper around the const reference. You could rewrite this function to do the same though. Although, it seems reasonably conservative to fail if there's an unexpected protocol passed in too